keps/sig-storage/20190204-scale-testing.md

```diff
@@ -0,0 +1,215 @@
+- TODO link to stateful pod SLI
+- https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/docs/design.md
+replaces:
+superseded-by:
+---
+
+# Volume Scale and Performance Testing Plan
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+   * [Goals](#goals)
+   * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+   * [WIP Test Environment](#wip-test-environment)
+   * [Test Cases](#test-cases)
+      * [Pod Startup](#pod-startup)
+      * [Pod Teardown](#pod-teardown)
+      * [WIP: PV Binding Tests](#wip-pv-binding-tests)
+      * [WIP: PV Provisioning Tests](#wip-pv-provisioning-tests)
+      * [WIP: PV Deletion Tests](#wip-pv-deletion-tests)
+* [Graduation Criteria](#graduation-criteria)
+   * [Phase 1](#phase-1)
+   * [Phase 2](#phase-2)
+* [Implementation History](#implementation-history)
+
+
+## Summary
+
+This KEP outlines a plan for testing scalability and performance of K8s storage components.
+
+## Motivation
+
+Adding storage scale and performance tests will help:
+* Understand the current scale limits of the Kuberentes storage system.
+* Set expectations (SLOs) for consumers of the Volume API.
+* Determine bottlenecks and influence which need addressing.
+
+### Goals
+
+* Measure the overhead of K8s components owned by sig-storage:
+  * K8s volume controllers
+  * Kubelet volume manager
+  * CSI sidecars
+* Stress various dimensions of volume operations to determine:
+  * Max volumes per pod
+  * Max volumes per node
+  * Max volumes per cluster
+* Test with the following volume types:
+  * EmptyDir
+  * Secret
+  * Configmap
+  * Downward API
+  * CSI mock driver
+  * TODO: Hostpath?
+  * TODO: Local?
+* Provide tests that vendors can easily run against their volume drivers.
+
+### Non-Goals
+
+* Test and measure storage provider’s drivers.
+
+## Proposal
+
+### Test Framework
+
+The tests should be developed and run in sig-scalability’s test
+[Cluster Loader framework](https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/README.md)
+and infrastructure.
+
+The framework already supports measurements for Pod startup latency and
+scheduling metrics. More measurements can be added to get finer-grained details
+for volume operations:
+
+* Gather [volume operation
+  metrics](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/metrics.go)
+  from kube-controller-manager
+  * Caveat: These metrics measure the time of a single operation attempt and not
+    the e2e time across retries. These metrics may not be suitable to measure
+    latency from a pod's perspective.
+* Measure time starting from pod scheduling to volume operation completion times
+  for provision, attach and mount. This requires changes in Kubernetes to add
+  timestamps in PersistentVolumeClaim, VolumeAttachment, and Pod statuses,
+  respectively.
+  * TODO: how to measure time reliably for deletion events, where the API object
+    deletion is the trigger?
```

Comment:
Good question, though I don't think it's a blocker.
We don't have any SLI for pod deletion and I've never heard high-level complaints that this is too slow.

Reply:
When using RWO volumes, pod deletion (and unmount/detach) time becomes more important because the next pod using the same volume cannot be started until the volume has been detached from the previous node.

Reply:
OK - that's a great point which I totally forgot about.
I will add a TODO to my proposal for the SLI to consider adding a deletion SLI because of that reason.

I still think it's not a hard blocker - I would treat it more as a second step.